File: server/brainteasers.js
const express = require("express");
const fs = require("fs");
const path = require("path");
const OpenAI = require("openai");
const crypto = require("crypto");

const router = express.Router();
const client = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

// ---------- simple per-IP rate limit ----------
const RATE = { windowMs: 60_000, max: 30 }; // 30 requests / minute / IP
const hits = new Map(); // ip -> [timestamps]
function rateLimit(req, res, next) {
  const ip = req.headers["x-forwarded-for"]?.toString().split(",")[0].trim() || req.socket.remoteAddress || "ip";
  const now = Date.now();
  const wStart = now - RATE.windowMs;
  const arr = (hits.get(ip) || []).filter(t => t > wStart);
  arr.push(now);
  hits.set(ip, arr);
  if (arr.length > RATE.max) return res.status(429).json({ error: "Too many requests, slow down." });
  next();
}
router.use(rateLimit);

// ---------- file cache helpers ----------
const mem = { byDate: new Map(), byId: new Map() };
const cacheDir = path.join(process.cwd(), ".cache");
const f = (dateStr) => path.join(cacheDir, `brainteasers-${dateStr}.json`);

function hashId(s) {
  return crypto.createHash("sha1").update(String(s)).digest("hex").slice(0, 16);
}

function loadFromDisk(dateStr) {
  try {
    const p = f(dateStr);
    if (fs.existsSync(p)) {
      const j = JSON.parse(fs.readFileSync(p, "utf8"));
      if (Array.isArray(j)) {
        mem.byDate.set(dateStr, j);
        j.forEach(t => mem.byId.set(t.id, { ...t, date: dateStr }));
        return j;
      }
    }
  } catch {}
  return null;
}

function saveToDisk(dateStr, items) {
  try {
    fs.writeFileSync(f(dateStr), JSON.stringify(items, null, 2));
  } catch {}
}

async function generateTeasersForDate(dateStr) {
  if (mem.byDate.has(dateStr)) return mem.byDate.get(dateStr);
  const disk = loadFromDisk(dateStr);
  if (disk) return disk;

  const prompt = `
You are a puzzle writer. Create exactly 5 short brainteasers/riddles for the date ${dateStr}.
Output strictly as JSON with this shape:
{ "teasers": [ { "prompt": "..." }, { "prompt": "..." }, { "prompt": "..." }, { "prompt": "..." }, { "prompt": "..." } ] }
Keep each prompt under 200 characters, self-contained, and without spoilers/answers.
`;

  let teasers = [];
  try {
    const completion = await client.chat.completions.create({
      model: "gpt-4o-mini",
      temperature: 0.8,
      messages: [
        { role: "system", content: "Return only valid JSON. No commentary." },
        { role: "user", content: prompt },
      ],
    });
    const text = completion.choices?.[0]?.message?.content || "{}";
    const parsed = JSON.parse(text);
    teasers = (parsed?.teasers || []).slice(0, 5);
  } catch {
    teasers = [
      { prompt: "What gets wetter the more it dries?" },
      { prompt: "Forward I am heavy, backward I am not. What am I?" },
      { prompt: "I have keys but open no locks. What am I?" },
      { prompt: "I speak without a mouth and hear without ears. What am I?" },
      { prompt: "Taken from a mine, trapped in wood, used by all. What am I?" },
    ];
  }

  const withIds = teasers.map(t => {
    const id = hashId(`${dateStr}:${t.prompt}`);
    return { id, prompt: String(t.prompt || "").trim() };
  });

  mem.byDate.set(dateStr, withIds);
  withIds.forEach(t => mem.byId.set(t.id, { ...t, date: dateStr }));
  saveToDisk(dateStr, withIds);
  return withIds;
}

// GET /api/brainteasers?date=YYYY-MM-DD
router.get("/brainteasers", async (req, res) => {
  try {
    const date = String(req.query.date || "").trim() || new Date().toISOString().slice(0,10);
    const teasers = await generateTeasersForDate(date);
    res.json({ teasers });
  } catch (e) {
    res.status(500).json({ error: e?.message || "Failed to fetch teasers" });
  }
});

// POST /api/brainteasers/check { id, answer }
router.post("/brainteasers/check", express.json(), async (req, res) => {
  try {
    const { id, answer } = req.body || {};
    if (!id || typeof answer !== "string")
      return res.status(400).json({ error: "id and answer required" });

    const item = mem.byId.get(String(id));
    if (!item) return res.status(404).json({ error: "Unknown teaser id" });

    const verifyPrompt = `
You are a strict riddle checker. The riddle is:
"${item.prompt}"

The user's answer is: "${answer}"

Reply ONLY valid JSON like:
{ "correct": true }  OR  { "correct": false }
Be lenient to common synonyms and spelling if the meaning matches.
`;

    let correct = false;
    try {
      const completion = await client.chat.completions.create({
        model: "gpt-4o-mini",
        temperature: 0,
        messages: [
          { role: "system", content: "Return only valid JSON. No commentary." },
          { role: "user", content: verifyPrompt },
        ],
      });
      const text = completion.choices?.[0]?.message?.content || "{}";
      const parsed = JSON.parse(text);
      correct = !!parsed?.correct;
    } catch { correct = false; }

    res.json({ correct, award: correct ? 5 : 0 });
  } catch (e) {
    res.status(500).json({ error: e?.message || "Check failed" });
  }
});

module.exports = router;

File: server/brain_teasers.py
import os, json, datetime, hashlib, re
from flask import Blueprint, request, jsonify
try:
    from openai import OpenAI
    _client = OpenAI(api_key=os.environ.get('OPENAI_API_KEY'))
except Exception:
    _client = None

bp = Blueprint('brain_teasers', __name__)

CACHE_DIR = os.path.join(os.path.dirname(__file__), 'cache')
os.makedirs(CACHE_DIR, exist_ok=True)

def _norm(s: str) -> str:
    return re.sub(r'[^a-z0-9]+', '', (s or '').strip().lower())

def _cache_path(date_str: str) -> str:
    return os.path.join(CACHE_DIR, f'brainteasers-{date_str}.json')

def _seed(date_str: str) -> str:
    salt = os.environ.get('BRAINTEASER_SALT', 'nova-salt')
    return hashlib.sha256(f'{date_str}-{salt}'.encode()).hexdigest()[:16]

def _generate(date_str: str):
    if _client is None:
        raise RuntimeError('OpenAI not available. Install openai and set OPENAI_API_KEY.')
    seed = _seed(date_str)
    prompt = (
        f'Generate exactly 5 short brain teasers for {date_str}.\n'
        'Return STRICT JSON: list of 5 objects with fields id, question, answer, category, difficulty (easy|medium|hard).\n'
        'Ensure a single best answer. Diverse types (logic, word, math, pattern).\n'
        f'Include seed {seed} inside id. No extra text, ONLY JSON.'
    )
    resp = _client.chat.completions.create(
        model='gpt-4o-mini',
        messages=[{'role':'user','content':prompt}],
        temperature=0.6,
    )
    text = resp.choices[0].message.content.strip()
    data = json.loads(text)
    if not isinstance(data, list) or len(data) != 5:
        raise ValueError('Model did not return 5 items.')
    for t in data:
        t['answer_norm'] = _norm(t.get('answer'))
    return data

@bp.route('/api/brainteasers', methods=['GET'])
def get_daily():
    date_str = request.args.get('date') or datetime.date.today().isoformat()
    path = _cache_path(date_str)
    if os.path.exists(path):
        with open(path, 'r') as f:
            teasers = json.load(f)
    else:
        teasers = _generate(date_str)
        with open(path, 'w') as f:
            json.dump(teasers, f)
    safe = [{k:v for k,v in t.items() if k not in ('answer_norm','answer')} for t in teasers]
    return jsonify({'date': date_str, 'teasers': safe})

@bp.route('/api/brainteasers/verify', methods=['POST'])
def verify():
    payload = request.get_json(force=True)
    date_str = payload.get('date') or datetime.date.today().isoformat()
    tid = payload.get('id')
    uans = payload.get('answer','')
    path = _cache_path(date_str)
    if not os.path.exists(path):
        return jsonify({'ok': False, 'reason':'not_ready'}), 400
    with open(path, 'r') as f:
        teasers = json.load(f)
    t = next((x for x in teasers if x.get('id') == tid), None)
    if not t:
        return jsonify({'ok': False, 'reason':'not_found'}), 404
    if _norm(uans) == t.get('answer_norm',''):
        return jsonify({'ok': True, 'match':'exact', 'correct_answer': t.get('answer')})
    if _client is None:
        return jsonify({'ok': False, 'reason':'no_semantic_check'}), 500
    judge = (
        f'Question: {t.get('question')}\n'
        f'Canonical answer: {t.get('answer')}\n'
        f'User answer: {uans}\n'
        'Reply ONLY JSON: {"correct": true|false}'
    )
    resp = _client.chat.completions.create(
        model='gpt-4o-mini',
        messages=[{'role':'user','content':judge}],
        temperature=0.0,
    )
    try:
        verdict = bool(json.loads(resp.choices[0].message.content.strip()).get('correct', False))
    except Exception:
        verdict = False
    return jsonify({
        'ok': verdict,
        'match': 'semantic' if verdict else 'none',
        'correct_answer': t.get('answer') if verdict else None
    })

__all__ = ['bp']
